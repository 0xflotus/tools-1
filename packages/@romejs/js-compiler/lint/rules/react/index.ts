import jsxNoCommentTextnodes from './jsxNoCommentTextnodes';

// Add transforms in alphabetical order.
export const reactLintTransforms = [jsxNoCommentTextnodes];
