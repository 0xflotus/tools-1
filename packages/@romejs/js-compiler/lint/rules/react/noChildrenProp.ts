/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {Path} from '@romejs/js-compiler';
import {AnyNode, JSXAttribute} from '@romejs/js-ast';
import {descriptions} from '@romejs/diagnostics';

export default {
  name: 'noChildrenProp',
  enter(path: Path): AnyNode {
    const {node} = path;

    if (
      (
        node.type === 'JSXElement' &&
        node.attributes.find(attribute => (attribute as JSXAttribute).name.name === 'children')
      ) || (
        node.type === 'CallExpression' &&
        node.callee.type === 'MemberExpression' &&
        node.callee.object.type === 'ReferenceIdentifier' &&
        node.callee.object.name === 'React' && 
        node.callee.property.value.type === 'Identifier' &&
        node.callee.property.value.name === 'createElement' &&
        node.arguments[1].type === 'ObjectExpression' &&
        node.arguments[1].properties.find(property => (
          property.type === 'ObjectProperty' &&
          property.key.value.type === 'Identifier' &&
          property.key.value.name === 'children'
        ))
      )
    ) {
      path.context.addNodeDiagnostic(
        node,
        descriptions.LINT.NO_CHILDREN_PROP,
      );
    }

    return node;
  },
};
