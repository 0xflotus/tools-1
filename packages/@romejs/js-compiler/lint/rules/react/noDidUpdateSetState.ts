/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { Path } from "@romejs/js-compiler";
import { AnyNode, ObjectProperty, AnyClassMember } from "@romejs/js-ast";
import { descriptions } from "@romejs/diagnostics";

function hasComponentDidUpdate(node: ObjectProperty | AnyClassMember): boolean {
  return (
    node.key.type === "StaticPropertyKey" &&
    node.key.value.type === "Identifier" &&
    node.key.value.name === "componentDidUpdate"
  );
}

function hasSetState(node: AnyNode): boolean {
  if (
    node.type === "ObjectProperty" &&
    node.value.type === "FunctionExpression"
  ) {
    if (node.value.body.type === "BlockStatement") {
      return node.value.body.body.some(
        statement =>
          statement.type === "ExpressionStatement" &&
          statement.expression.type === "CallExpression" &&
          statement.expression.callee.type === "MemberExpression" &&
          statement.expression.callee.property.value.type === "Identifier" &&
          statement.expression.callee.property.value.name === "setState"
      );
    }
  }
  if (node.type === "ClassMethod") {
    if (node.body.type === "BlockStatement") {
      return node.body.body.some(
        statement =>
          statement.type === "ExpressionStatement" &&
          statement.expression.type === "CallExpression" &&
          statement.expression.callee.type === "MemberExpression" &&
          statement.expression.callee.property.value.type === "Identifier" &&
          statement.expression.callee.property.value.name === "setState"
      );
    }
  }
  return false;
}

function hasSetStateInComponentDidUpdate(node: AnyNode): boolean {
  if (node.type === "ObjectExpression") {
    return node.properties.some(
      prop =>
        prop.type === "ObjectProperty" &&
        hasComponentDidUpdate(prop) &&
        hasSetState(prop)
    );
  }

  if (node.type === "ClassDeclaration") {
    return node.meta.body.some(
      method => hasComponentDidUpdate(method) && hasSetState(method)
    );
  }
  return false;
}

export default {
  name: "noDidUpdateSetState",
  enter(path: Path): AnyNode {
    const { node } = path;

    if (hasSetStateInComponentDidUpdate(node)) {
      path.context.addNodeDiagnostic(
        node,
        descriptions.LINT.NO_DID_UPDATE_SET_STATE
      );
    }
    return node;
  }
};
