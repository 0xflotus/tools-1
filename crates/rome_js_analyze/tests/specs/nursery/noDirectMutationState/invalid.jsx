import { Component } from 'react';

class A extends Component {
    render() {
        this.state = 1;
    }
}