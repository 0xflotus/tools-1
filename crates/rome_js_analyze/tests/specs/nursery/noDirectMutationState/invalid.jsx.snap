---
source: crates/rome_js_analyze/tests/spec_tests.rs
expression: invalid.jsx
---
# Input
```js
import { Component } from 'react';

class A extends Component {
    render() {
        this.state = 1;
    }
}
```

# Diagnostics
```
invalid.jsx:5:9 lint/nursery/noDirectMutationState ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ! Do not mutate React components state directly
  
    3 │ class A extends Component {
    4 │     render() {
  > 5 │         this.state = 1;
      │         ^^^^^^^^^^
    6 │     }
    7 │ }
  
  i See React dpcumentation
  

```


